src/bigquery.ts

```diff
@@ -1053,18 +1127,52 @@ export class BigQuery extends common.Service {
         query.queryParameters = [];
 
         // tslint:disable-next-line forin
-        for (const namedParamater in query.params) {
-          const value = query.params[namedParamater];
-          const queryParameter = BigQuery.valueToQueryParameter_(value);
-          queryParameter.name = namedParamater;
+        for (const namedParameter in query.params) {
+          const value = query.params[namedParameter];
+          let queryParameter;
+
+          if (query.types) {
+            if (!is.object(query.types)) {
+              throw new Error(
+                'Provided types must match the value type passed to `params`'
+              );
+            }
+            queryParameter = BigQuery.valueToQueryParameter_(
+              value,
+              query.types[namedParameter]
```

Comment:
Would it be throwable if this returned undefined, e.g. a user provided the types object, but only had types for some (or none) of the parameters? (Same question for the array version of code beneath this section)

Reply:
I'll add this.